### DISCUSSION TITLE
Mirai Re-design

### DISCUSSION THREAD

**Discussion by shikokuchuo** [r-lib/mirai#38] [Ideas]
Which was always in the works even prior to the 0.8.0 release. Inviting @wlandau.

This uses synchronisation primitives from the NNG library newly-exposed in `nanonext` (to be 0.8.1). We should confirm we have all we need for `mirai` before we release `nanonext`.

**Comment by shikokuchuo** [comment]
9a0f2a6 begins the journey.

Re-writes `server()` switching from a blocking receive to a waiting on a condition variable. This condition variable is signalled either by an async recv or a pipe notification event (on pipe removal). When this happens, we simply check to see if the aio is unresolved - if it is then it is for sure a pipe event and we exit.

This leads to much safer behaviour. For example we can now have a local dispatcher with daemons, kill the dispatcher process and the daemons will all exit (was not the case previously).

Also no need to send any exit signals!! A bit less magic, but bigger overall magic!

Another side effect is that the server process exits gracefully and silently in all scenarios now rather than through error.

@wlandau when you have the chance to test the new behaviour, it might allow you to remove some of your safeguards in `crew`. I am doing this piecemeal, but when I have finished what I already have planned, we can look at what else is needed to make the overall integration with `crew` better.

**Reply by wlandau** [reply]
This is so exciting! I cannot thank you enough for continuing to improve mirai. Your work has moved my own work forward so much.

The improved server exit behavior is really encouraging. Would it be possible to do the same for the dispatcher, i.e. exit whenever the client process exits? With the current CRAN mirai, if I call daemons(dispatcher = TRUE) from RStudio (haven't tested this in the terminal) on my local MacBook and then restart R, my local R session crashes and the dispatcher keeps running. That makes me wonder if there is something about the client/dispatcher relationship that can be repurposed as an extra safeguard. I also wonder about the feasibility of a more robust replacement for the one-second lifespan buffer we talked about, plus maybe a long configurable idle timeout if it would prevent it from running amok on shared systems. As you may have noticed, I have been having trouble trusting that the dispatcher will exit on its own.

**Reply by shikokuchuo** [reply]
All of these will be addressed in good time. The dispatcher will exit in the same way as servers, so no need to worry especially about a dispatcher process. Both dispatchers and servers will exit if the main session is closed (forcibly or not).

**Reply by wlandau** [reply]
Thank you, I am very much looking forward to that! On my end, I think the last known issue is https://github.com/wlandau/crew/issues/51, and at the moment I have no idea whether it is caused by `crew` or `mirai`.

**Reply by shikokuchuo** [reply]
wait a couple more days for the non-polling dispatcher. it will be a game-changer. congrats on `crew` 0.0.4 on CRAN! It's been a real pleasure working together on this.

**Reply by wlandau** [reply]
Such a game-changer. As I write documentation on custom worker launchers in `crew`, I am realizing how liberating it will be to rely entirely on the `mirai` dispatcher in order to terminate servers. If `crew` does not have to manually terminate its workers, then writing these plugins will be a whole lot easier and more efficient, and more people from the community will be able to participate.

**Reply by wlandau** [reply]
At the very least, I can make manual worker termination optional. Some authors might want to add it in the case when startup times are  really long, and some authors may not be able to add one, depending on the platform.

**Comment by shikokuchuo** [comment]
With shikokuchuo/nanonext#[5073346](https://github.com/shikokuchuo/nanonext/commit/5073346d587918c1548e59cfe7e8739aab94805c) we should now be up to the task.

Pipe events (i.e. disconnects) can be easily distinguished as they can be set to trip a flag. The `wait()` will then return FALSE instead of TRUE.

This allows us to write clean code such as:
```r
wait(cv) || break
```
And we don't need to check for unresolved status any more.

**Reply by shikokuchuo** [reply]
It's almost comical how I can keep adding to my arsenal. [e726170](https://github.com/shikokuchuo/nanonext/commit/e726170f0ca42791430771b785c34f886f8b6445) is not enough...

**Reply by shikokuchuo** [reply]
turns out [fc08c77](https://github.com/shikokuchuo/nanonext/commit/fc08c77b9bb2ad85d452b6b8a0d238a58b38b0b8) was needed rather than the previous commit. This is now enough :)

**Comment by shikokuchuo** [comment]
This has come in one drop as lots of stuff tied together, but the non-polling dispatcher is complete in [3891ddf](https://github.com/shikokuchuo/mirai/commit/3891ddfca4d763498d0f66a1bfa135c4e61448fc).

Has gone through a fair amount of stress testing and passes well.

This throughput test is now instantaneous:
https://github.com/wlandau/crew/issues/41#issue-1623044321

My original test which spurred all of this, now returns an overhead of an incredibly miniscule 13 ms on my laptop!!
https://github.com/shikokuchuo/mirai/discussions/18#discussioncomment-4982041

In no particular order:

'pollfreqh' and 'pollfreql' arguments removed as have no more meaning. 'exitlinger' removed from dispatcher as no longer applicable there. These should be backwards compatible as extra arguments should get lost and ignored in the '...'.

No minimum socket lifespan needed. There is now a little ack exchange between client and dispatcher to guard against hanging processes such as: https://github.com/wlandau/crew/issues/41#issue-1623044321. Manually-launched dispatchers are still possible though and does not have this to allow normal operation.

dispatcher and server will now exit gracefully when parent process exits/crashes etc. (whenever the connection is dropped). This should help minimise orphan processes.

Self-repairing local daemons (when auto-launched) was a nice proof of concept, but removed to reduce complexity and enhance overall logic flow at dispatcher. Would have been rarely needed for the local cases.

'exitlinger' applicable now only to `server()` when time or task outs occur to ensure data gets sent back, not when client/dispatcher exits. Default is now set at 1000ms - only needs to be set wider if returning multi-GB objects. It has not been possible to eliminate this entirely for reasons I will detail in a new issue.

`dispatcher()` now uses cvs and `pipe-notify()` to capture all server connects and disconnects - and stats on instances are now accurate even if servers just connect and disconnect repeatedly. It no longer uses the NNG stats interface which we found to not really be real-time. And a nice side effect is getting the connection status is now 1-2 orders of magnitude faster!

**Reply by wlandau** [reply]
All these changes are wonderful! `crew` is so much faster, and the dispatcher is snappy and trustworthy in terms of exits. I did submit a couple bug reports, but I can work around those in `crew`. https://github.com/wlandau/crew/issues/51 still persists, but that may be something in `crew` in not `mirai`, I still do not know.

**Reply by shikokuchuo** [reply]
> All these changes are wonderful! `crew` is so much faster, and the dispatcher is snappy and trustworthy in terms of exits.

Thanks! I knew what I wanted conceptually, but couldn't be certain if it was going to be actually feasible. In the end it turned out to be fine, although involving quite a bit of engineering at {nanonext}. A straight wrapping of NNG's functionality would not have been up to the task - the implementation is highly customised to our needs!
